sel/src/Sel/PublicKey/AuthenticatedEncryption.hs

```diff
@@ -0,0 +1,494 @@
+hashToBinary (Hash messageLength fPtr) =
+  BS.fromForeignPtr0
+    (Foreign.castForeignPtr fPtr)
+    (fromIntegral messageLength + fromIntegral cryptoBoxMACBytes)
+
+-- | Create an authenticated hash from a message, a secret key,
+-- and a one-time cryptographic nonce that must never be re-used with the same
+-- secret key to encrypt another message.
+--
+-- @since 0.0.1.0
+encrypt
+  :: StrictByteString
+  -- ^ Message to encrypt.
+  -> SecretKey
+  -- ^ Secret key of the sender
+  -> PublicKey
+  -- ^ Public key of the recipient
+  -> IO (Nonce, Hash)
+encrypt message (SecretKey secretKeyForeignPtr) (PublicKey publicKeyForeignPtr) =
+  BS.unsafeUseAsCStringLen message $ \(cString, cStringLen) -> do
+    (Nonce nonceForeignPtr) <- newNonce
+    hashForeignPtr <- Foreign.mallocForeignPtrBytes (cStringLen + fromIntegral cryptoBoxMACBytes)
+    Foreign.withForeignPtr hashForeignPtr $ \hashPtr ->
+      Foreign.withForeignPtr secretKeyForeignPtr $ \secretKeyPtr ->
+        Foreign.withForeignPtr publicKeyForeignPtr $ \publicKeyPtr ->
+          Foreign.withForeignPtr nonceForeignPtr $ \noncePtr -> do
+            void $
+              cryptoBoxEasy
+                hashPtr
+                (Foreign.castPtr @CChar @CUChar cString)
+                (fromIntegral @Int @CULLong cStringLen)
+                noncePtr
+                publicKeyPtr
+                secretKeyPtr
+            let hash = Hash (fromIntegral @Int @CULLong cStringLen) hashForeignPtr
+            pure (Nonce nonceForeignPtr, hash)
+
+-- | Decrypt a hashed and authenticated message with the shared secret key and the one-time cryptographic nonce.
+--
+-- @since 0.0.1.0
+decrypt
+  :: Hash
+  -- ^ Encrypted message you want to decrypt.
+  -> SecretKey
+  -- ^ Secret key of the recipient.
+  -> PublicKey
+  -- ^ Public key of the sender.
+  -> Nonce
+  -- ^ Nonce used for encrypting the original message.
+  -> Maybe StrictByteString
+decrypt
+  Hash{messageLength, hashForeignPtr}
+  (SecretKey secretKeyForeignPtr)
+  (PublicKey publicKeyForeignPtr)
+  (Nonce nonceForeignPtr) = unsafeDupablePerformIO $ do
+    messagePtr <- Foreign.mallocBytes (fromIntegral @CULLong @Int messageLength)
+    Foreign.withForeignPtr hashForeignPtr $ \hashPtr ->
+      Foreign.withForeignPtr secretKeyForeignPtr $ \secretKeyPtr ->
+        Foreign.withForeignPtr publicKeyForeignPtr $ \publicKeyPtr ->
+          Foreign.withForeignPtr nonceForeignPtr $ \noncePtr -> do
+            result <-
+              cryptoBoxOpenEasy
+                messagePtr
+                hashPtr
+                (messageLength + fromIntegral cryptoBoxMACBytes)
+                noncePtr
+                publicKeyPtr
+                secretKeyPtr
+            case result of
+              (-1) -> pure Nothing
+              _ -> do
+                bsPtr <- Foreign.mallocBytes (fromIntegral messageLength)
+                memcpy bsPtr (Foreign.castPtr messagePtr) (fromIntegral messageLength)
```

Comment:
Is this copy needed? AFAICS `messagePtr` can be used directly if one changes `CChar` to `CUChar` below.

Reply:
Taking into account the dangerous nature of `unsafePackMallocCStringLen` I thought it would be safer. But also I have much to learn in terms of determining when boundaries are violated.